src/framework/jobState/LocalStepJobState/LocalStepJobState.ts

```diff
@@ -0,0 +1,97 @@
+import { Promisable } from 'type-fest';
+
+import { Entity, Relationship } from '../../types';
+import { JobState, GraphObjectFilter, GraphObjectIteratee } from '../types';
+
+import { flushDataToDisk } from './flushDataToDisk';
+
+export const GRAPH_OBJECT_BUFFER_THRESHOLD = 500; // arbitrarily selected, subject to tuning
+
+interface LocalStepJobStateInput {
+  step: string;
+  cacheDirectory?: string;
+}
+
+export class LocalStepJobState implements JobState {
+  readonly cacheDirectory?: string;
+  readonly step: string;
+
+  entities: Entity[];
+  relationships: Relationship[];
+
+  constructor({ step, cacheDirectory }: LocalStepJobStateInput) {
+    this.step = step;
+    this.cacheDirectory = cacheDirectory;
+
+    this.entities = [];
+    this.relationships = [];
+  }
+
+  async addEntities(newEntities: Entity[]) {
+    this.entities = this.entities.concat(newEntities);
+
+    if (this.entities.length >= GRAPH_OBJECT_BUFFER_THRESHOLD) {
+      await this.flushEntitiesToDisk();
+    }
+  }
+
+  async addRelationships(newRelationships: Relationship[]) {
+    this.relationships = this.relationships.concat(newRelationships);
+
+    if (this.relationships.length >= GRAPH_OBJECT_BUFFER_THRESHOLD) {
+      await this.flushRelationshipsToDisk();
+    }
+  }
```

Comment:
Any reason to care about using the sum, such as `entities + relationships >= GRAPH_OBJECT_BUFFER_THRESHOLD`? Without it, the memory could grow to twice as much as that number constant indicates, I think.

Reply:
I was thinking that the worse that could happen would be having 1000 entities or relationships in memory, which doesn't seem like too much to me.

Reply:
Agreed, it's nothing really. I think what you have is just fine. I should have written, it was surprising to me when I read through the code that there could be more than `GRAPH_OBJECT_BUFFER_THRESHOLD` objects not flushed.